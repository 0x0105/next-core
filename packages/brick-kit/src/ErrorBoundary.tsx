import React from "react";
import {
  withTranslation,
  WithTranslation,
  // `WithTranslationProps` is required for api-extractor.
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  WithTranslationProps,
} from "react-i18next";
import { httpErrorToString } from "./handleHttpError";
import { NS_BRICK_KIT, K } from "./i18n/constants";

interface ErrorBoundaryState {
  error: any;
}

// Ref https://reactjs.org/docs/error-boundaries.html
class LegacyErrorBoundary extends React.Component<
  WithTranslation,
  ErrorBoundaryState
> {
  constructor(props: WithTranslation) {
    super(props);
    this.state = { error: null };
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    // Update state so the next render will show the fallback UI.
    return { error };
  }

  /* componentDidCatch(error, info) {
    // You can also log the error to an error reporting service
    logErrorToMyService(error, info);
  } */

  render(): React.ReactNode {
    if (this.state.error) {
      // You can render any custom fallback UI
      return (
        <div data-testid="error-boundary">
          <h3>{this.props.t(K.SOMETHING_WENT_WRONG)}</h3>
          <p>{httpErrorToString(this.state.error)}</p>
        </div>
      );
    }

    return this.props.children;
  }
}

/** @internal */
export const ErrorBoundary = withTranslation(NS_BRICK_KIT)(LegacyErrorBoundary);
