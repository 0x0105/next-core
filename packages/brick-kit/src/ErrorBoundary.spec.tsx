import React from "react";
import { shallow } from "enzyme";
import { ErrorBoundary } from "./ErrorBoundary";

function Child(): any {
  // this is just a placeholder
  return null;
}

describe("ErrorBoundary", () => {
  it("should show errors", async () => {
    const wrapper = shallow(
      <ErrorBoundary>
        <Child />
      </ErrorBoundary>
    );
    expect(wrapper).toMatchSnapshot();
    const error = new Error("oops");
    wrapper.find(Child).simulateError(error);
    expect(wrapper).toMatchSnapshot();
  });
});
