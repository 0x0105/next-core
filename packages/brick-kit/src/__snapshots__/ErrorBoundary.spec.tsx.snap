// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should show errors 1`] = `<Child />`;

exports[`ErrorBoundary should show errors 2`] = `
<div>
  <h3>
    SOMETHING_WENT_WRONG
  </h3>
  <p>
    Error: oops
  </p>
</div>
`;
